File: AoC2024Day10.cpp
import std.core;

template<typename T>
struct Coord
{
	constexpr bool operator <  (const Coord& p) const { if (x < p.x) return true; else if (p.x < x) return false; else return y < p.y; }
	constexpr Coord operator + (const Coord& p) const { return Coord(x + p.x, y + p.y); }
	
	template<typename U, std::enable_if_t<std::is_integral_v<T>, bool> = true>
	bool Within(const U& grid) const { return x >= 0 && y >= 0 && x < grid.size() && y < grid[x].size(); }
	template<typename U, std::enable_if_t<std::is_integral_v<T>, bool> = true>
	const auto& GetAt(const U& grid) const { return grid[x][y]; }

	T x = {}, y = {};
};
using Point = Coord<int>;
constexpr std::array<Point, 4> directions = { Point(1, 0), Point(0, 1), Point(-1, 0), Point(0,-1) };

template <typename T>
class y_combinator {
	T lambda;
public:
	constexpr y_combinator(T&& t) : lambda(std::forward<T>(t)) {}
	template <typename...Args>
	constexpr decltype(auto) operator()(Args&&...args) const { return lambda(std::move(*this), std::forward<Args>(args)...); }
};

int main(int argc, char* argv[])
{
	auto ChronoStart = std::chrono::high_resolution_clock::now();
	if (argc < 2)
	{
		std::cout << "Usage: AoC24Day10.exe inputFilename\n";
		return -1;
	}
	std::ifstream in(argv[1]);
	if (!in)
	{
		std::cout << std::format("Could not open inputFilename {}\n", argv[1]);
		return -1;
	}

	int part1 = 0, part2 = 0;
	std::string line;
	std::vector<std::vector<int>> grid;
	
	while (in >> line)
	{
		std::vector<int> v;
		for (char c : line)
			v.push_back(c-'0');
		grid.push_back(std::move(v));
	}

	auto PathCount = y_combinator([&grid](auto && PathCount, Point p, int step) -> int
	{
		if (step++ >= 9)
			return 1;
		int result = 0;
		for (Point d : directions)
			if (Point np = p + d; np.Within(grid) && np.GetAt(grid) == step)
				result += PathCount(np, step);

		return result;
	});

	for (int i = 0; i < grid.size(); ++i)
		for (int j = 0; j < grid.front().size(); ++j)
			if (grid[i][j] == 0)
			{
				std::set<Point> cur, next;
				cur.insert({i, j});
				int iter = 1;
				while (iter < 10 && !cur.empty())
				{
					for (Point p : cur)
						for (Point d : directions)
							if (Point np = p + d; np.Within(grid) && np.GetAt(grid) == iter)
								next.insert(np);
					++iter;
					std::swap(cur, next);
					next.clear();
				}

				part1 += cur.size();
				part2 += PathCount(Point{i, j}, 0);
			}

	std::cout << std::format("Part 1: {}\nPart 2: {}\n", part1, part2);
	std::cout << std::format("Duration: {}\n", std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now() - ChronoStart));
}

File: AoC2024Day10.fs
open System.IO

let ToGrid (s:string) =
    s.ToCharArray() |> Array.map(fun c -> (int)c - (int)'0')

let directions = [|(1,0); (0,1); (-1,0); (0,-1)|]
    
[<EntryPoint>]
let main argv =
    let grid = File.ReadAllLines(argv[0]) |> Array.map ToGrid
    let MaxX, MaxY = Array.length grid, Array.head grid |> Array.length
    let Within (x,y) = x>=0 && y>=0 && x<MaxX && y<MaxY

    let rec CountTrails (x,y) level =
        if not (Within (x,y)) || grid[x][y] <> level then []
        else if level=9 then [(x,y)]
             else Array.fold(fun points (dx,dy) -> List.append points (CountTrails (x+dx,y+dy) (level+1))) [] directions

    let rec CountPaths (x,y) level =
        if not (Within (x,y)) || grid[x][y] <> level then 0
        else if level=9 then 1
            else Array.fold(fun t (dx,dy) -> t + (CountPaths (x+dx, y+dy) (level+1))) 0 directions

    let Process f = Array.fold(fun total (x, line) -> total + Array.fold(fun t (y,v) -> t + (f (x,y))) 0 (Array.indexed line)) 0 (Array.indexed grid)
    let Part1 pos = List.distinct (CountTrails pos 0) |> List.length
    let Part2 pos = CountPaths pos 0
    printfn("Part 1: %A") <| Process Part1
    printfn("Part 2: %A") <| Process Part2
    0